# BOMA
Brain and Organoid Manifold Alignment

## Summary
3D organoids cultured from induced human pluripotent stem cells (hiPSCs) are useful models for understanding the longitude genomical dynamics during human brain development, however, the genomic differences between the brains and organoids are not well established. Here, we applied manifold alignment (MA) to comparatively analyze both the bulk and single-cell transcriptomic measurements across time among human brains, human organoids, Non-Human Primates (NHP) organoids within published literatures. Our MA method discovered similar/different developmental patterns across the systems as reported before but in a much simpler manner. More importantly, the developmental trajectories generated by MA served as references for comparing single-cell transcriptomes between human brains and organoids, which allowed us to generate a database summarizing gene expression dynamics across time and cell-types. We experimentally validated some of the discrepancies observed between the two systems. Finally, we leverage our method to align single-cell transcriptomic measurements of human brains (n=57) and organoids (n=35) by integrating data from multiple recent publications. This alignment results demonstrated the capability of our method to integratively and comparatively analyze datasets generated from various experimental sources, which would be useful for understanding brains versus organoids differences and guiding the future organoids protocol design.

## Flow chart
As shown in the flowchart below,here we introduce the comparative analysis pipeline for aligning human and organoid developmental transcriptomic datasets.manifold alignment mainly contains two steps: global alignment and local alignment.


<div align=center><img width="1000" height="500" src="https://github.com/daifengwanglab/BOMA/blob/main/image/1.png"/></div>

## Software Requirements
The following analysis is based on R 4.0. 
Users should install the following packages prior to using the code:
```R
install.packages(c('rdist','SCORPIUS','Rmagic','pheatmap','Linnorm','reshape2','reticulate','plyr','RColorBrewer','stringr','limma','ManiNetCluster','ComplexHeatmap','circlize'))
# import ManiNetCluster form "https://github.com/daifengwanglab/ManiNetCluster"
```
Besides,some functions for processing data are also needed for our project.
```R
source('../src/func.r')
```

## Online webApp
An online web App to perform manifold alignment with user provided RNA-seq matrices and correspondence matrix can be found [here](https://github.com/Oafish1/ManiNetCluster-Visualization).


## Demo for aligning bulk RNA-seq between human brain and organoid
### Data description
we applied BOMA to align developmental gene expression data between brain samples from BrainSpan (Li M, et. al., Science, 2018) (N=460 from 16 human brain regions) with organoid samples from a recent long-term cultured ‘human cortical spheroid (hCS)’ organoid bulk RNA-seq dataset13 (N=62). BrainSpan covers in-vivo human brain development from 8PCW until 40 years old, including 28 time points. The organoid dataset covers the in vitro 3D culturing of human cortical spheroid (hCS) from 25 days up until 2 years,including 12 time points.  
Below shows the whole pipeline for aligning the two datasets, while the alignment (Step 3) took less 0.3s to complemet.




### Step 1: Data preprocessing
load and pre-process the gene expression matrices 
```R
#rdata1, rdata2 are the count matrices of the two datasets.
#meta1, meta2 contains the meta information for two datasets
load('bulk_start.RData')
rdata1 = rdata_human; meta1=meta_human
rdata2 = rdata_organoid; meta2=meta_organoid
# create a data set that only contains human information
regions = unique(meta1$Region)
form.data1=rdata1[,meta1$Species=='Human' & meta1$Region %in% regions]
form.meta1=meta1[meta1$Species=='Human' & meta1$Region %in% regions,]
form.data2=rdata2
form.meta2=meta2
# add a column representing the observation time to each meta data set.
form.meta1$time=form.meta1$Period   
form.meta2$time=form.meta2$Days
```


### Step 2: Feature(Gene) selection
To focus our analysis within brain development only, we identified 1,533 genes (see paper) most related to human brain development. We took the intersection of genes that are differentially expressed in human and organoid cells, and then intersected with these 1,533 genes for feature selection.  
We use 'limma' package to perform differential analysis at each time to find differentially expressed genes.
```R
# The specific information of a function is in "func.r" 
deg_list1 = DE.list(form.data1,form.meta1)
deg_list2 = DE.list(form.data2,form.meta2)
```
```R
sel.genes = intersect(intersect(deg_list1,deg_list2),unique(all.rec$gene))  # Select genes of interest
sel.data1 = form.data1[row.names(form.data1) %in% sel.genes,]  # Select expression based genes
sel.data1 = sel.data1[!duplicated(row.names(sel.data1)),] 
sel.meta1=form.meta1  #on human data
sel.data2 = form.data2[row.names(form.data2) %in% sel.genes,] 
sel.data2 = sel.data2[!duplicated(row.names(sel.data2)),] 
sel.meta2=form.meta2  #on organoid data
```
```R
# perform logarithmic transformation 
exp1 = sel.data1[order(row.names(sel.data1)),order(sel.meta1$time)];
sel.meta1=sel.meta1[order(sel.meta1$time),]
exp2 = sel.data2[order(row.names(sel.data2)),order(sel.meta2$time)];
sel.meta2=sel.meta2[order(sel.meta2$time),]
# reordering processing
ps.mat1 = t(exp1);ps.time1=sel.meta1$time
ps.mat2 = t(exp2);ps.time2=sel.meta2$time
```
### Step 3: BOMA Manifold Alignment
Manifold learning to uncover and match local and non-linear structures among datasets.
```R
algn_res = runMSMA_dtw(ps.mat1,ps.mat2)
df2 = algn_res[[3]]
df2$time = c(sel.meta1$time,sel.meta2$time)  # Add the time information to the aligned matrix
```


### Step 4: Downstream analysis and visualization of alignment results
Below, we show 3 ways for analyzing BOMA results.
1. Heatmap for visualizing pairwise similarity between aligned samples. 
```R
# calculate pairwise distances between cells after MSMA
pair_dist = apply(df2[df2$data=='sample1',c(3:5)],1,function(x) {
        d = apply(df2[df2$data=='sample2',c(3:5)],1,function(y) eu.dist(x,y))
})

row.names(pair_dist)=row.names(ps.mat2)
colnames(pair_dist)=row.names(ps.mat1)
sim_dist = 1/(1+pair_dist)

# hmtp3-2nd align
cols1 = c(brewer.pal(9,'YlGnBu')[c(2:9)],brewer.pal(11,'BrBG')[c(8:11)]);names(cols1) = unique(ps.time1)
annot1=rowAnnotation(H_time=factor(ps.time1,levels=unique(ps.time1)),col=list(H_time=cols1))
cols2 = c(brewer.pal(9,'YlOrRd')[c(2:9)],brewer.pal(11,'BrBG')[rev(1:4)]);names(cols2) = unique(ps.time2)
annot2=columnAnnotation(O_time=factor(ps.time2,levels=unique(ps.time2)),col=list(O_time=cols2))

sim_mat = t(sim_dist)
pdf('hmtp.pdf')
Heatmap(sim_mat,name='human_vs_humanOrg',show_row_names=F,show_column_names=F,cluster_rows=F,cluster_columns=F,
left_annotation=annot1,top_annotation=annot2,col=colorRamp2(c(0.8,max(sim_dist)),c('white','red')))
dev.off()
```
<div align=center><img width="500" height="500" src="https://github.com/daifengwanglab/BOMA/blob/main/image/hmtp_00.png"/></div>

2. Embedding of human brain and organoid samples in the aligned manifold space. Brain samples were colored by brain developmental stages. Organoid samples were colored by cultured days.
```R
# plot 3D trajectors
#scatter plot for human brain samples
pdf('3D1.pdf')
time.cols1 = colorRampPalette(brewer.pal(n=9,'Greens'))(12)

res = data.frame(df2[df2$data=='sample1',])
res0 = data.frame(df2)
library(plot3D)
s3d<-scatter3D(x=res[,3],y=res[,4],z=res[,5],colvar=as.numeric(mapvalues(res$time,names(table(res$time)),c(2:13))),col = c(time.cols1),pch=c(16,17)[as.numeric(as.factor(res$data))],colkey=F,theta = 300, phi = 30,cex=2,
xlim=c(min(res0$Val0),max(res0$Val0)),
ylim=c(min(res0$Val1),max(res0$Val1)),
zlim=c(min(res0$Val2),max(res0$Val2)))
legend("top", legend = levels(as.factor(res$data)), pch = c(16, 17),inset = -0.1, xpd = TRUE, horiz = TRUE)
legend("right", legend = levels(as.factor(res$time)), col = c(time.cols1),pch=16,inset =-0.05, xpd = TRUE, horiz = F,cex=1.2)
dev.off()
```

```R
#scatter plot for organoid samples
time.cols2 = colorRampPalette(brewer.pal(9, "Blues"))(12) #for organoid

res = data.frame(df2[df2$data=='sample2',])
#res = data.frame(df2)
res0 = data.frame(df2)
library(plot3D)
pdf('3D2.pdf')
s3d<-scatter3D(x=res[,3],y=res[,4],z=res[,5],pch=24,col='gray',bg=time.cols2[as.numeric(mapvalues(res$time,names(table(res$time)),c(1:12)))],colkey=F,theta = 300, phi = 30,cex=2,lwd=1,
xlim=c(min(res0$Val0),max(res0$Val0)),
ylim=c(min(res0$Val1),max(res0$Val1)),
zlim=c(min(res0$Val2),max(res0$Val2)))
legend("top", legend = levels(as.factor(res$data)), pch = c(16, 17),inset = -0.1, xpd = TRUE, horiz = TRUE)

legend("right", legend = levels(as.factor(res$time)), col = c(time.cols2),pch=16,inset = -0.05, xpd = TRUE, horiz = F,cex=1.2)
dev.off()
```
<img width="400" height="400" src="https://github.com/daifengwanglab/BOMA/blob/main/image/3D1_300_30_00.png"/></div><img width="400" height="400" src="https://github.com/daifengwanglab/BOMA/blob/main/image/3D1_200_60-2_00.png"/></div>

<img width="400" height="400" src="https://github.com/daifengwanglab/BOMA/blob/main/image/3D2_300_30_00.png"/></div><img width="400" height="400" src="https://github.com/daifengwanglab/BOMA/blob/main/image/3D2_200_60_00.png"/></div>


3. Corrplot to show the averaged similarity accross time points.
```R
library(corrplot)
sim_avg = matrix(0,nrow=length(unique(ps.time1)),ncol=length(unique(ps.time2)))

i = 0
for(t1 in unique(ps.time1)) {
        i = i+1
        j = 0
        for (t2 in unique(ps.time2)) {
                j = j+1
                sim_tmp = sim_mat[ps.time1==t1,ps.time2==t2]
                avg_tmp = mean(sim_tmp)
                sim_avg[i,j] = avg_tmp
        }
}

sim_avg = t(sim_avg)
colnames(sim_avg) = unique(ps.time1)
row.names(sim_avg) = unique(ps.time2)

library(RColorBrewer)
pdf('corr.plot.bulk.human.vs.org1.pdf')
corrplot(sim_avg,col=rev(colorRampPalette(brewer.pal(9,'PRGn'))(200)),cl.lim=c(0.7,1),is.corr=F,tl.col="black")
dev.off()
```
<div align=center><img width="500" height="500" src="https://github.com/daifengwanglab/BOMA/blob/main/image/corr.plot.bulk.human.vs.org1_00.png"/></div>

